File: go/program/ops/opFunctionDefinition.go
package ops

import (
	"github.com/2manyvcos/jpl/go/definition"
	"github.com/2manyvcos/jpl/go/jpl"
	"github.com/2manyvcos/jpl/go/library"
)

type opFunctionDefinition struct{}

// { argNames: [string], pipe: [op] }
func (opFunctionDefinition) OP(runtime jpl.JPLRuntime, input any, params definition.JPLInstructionParams, scope jpl.JPLRuntimeScope, next jpl.JPLScopedPiper) ([]any, jpl.JPLError) {
	return next.Pipe(library.ScopedFunction(params.ArgNames, params.Pipe, scope), scope)
}

// { argNames: [string], pipe: function }
func (opFunctionDefinition) Map(runtime jpl.JPLRuntime, params jpl.JPLInstructionParams) (definition.JPLInstructionParams, jpl.JPLError) {
	return definition.JPLInstructionParams{
		ArgNames: params.ArgNames,
		Pipe:     call(params.Pipe),
	}, nil
}

File: js/lib/program/ops/opFunctionDefinition.ts
import type { JPLInstructionParams } from '@/library/definition';
import type JPLRuntime from '@/runtime';
import { scopedFunction, type JPLRuntimeScope } from '../../library';
import type { UserJPLInstructionParams } from '../params';
import { call } from './utils';

export default {
  /** { argNames: [string], pipe: [op] } */
  op(
    _runtime: JPLRuntime,
    _input: unknown,
    params: JPLInstructionParams,
    scope: JPLRuntimeScope,
    next: (
      output: unknown,
      scope: JPLRuntimeScope,
    ) => Promise<unknown[]> | unknown[],
  ): Promise<unknown[]> | unknown[] {
    return next(
      scopedFunction(params.argNames ?? [], params.pipe ?? [], scope),
      scope,
    );
  },

  /** { argNames: [string], pipe: function } */
  map(
    runtime: JPLRuntime,
    params: UserJPLInstructionParams,
  ): JPLInstructionParams {
    return {
      argNames: runtime.muxOne([params.argNames!], (entry) =>
        runtime.assertType(entry, 'string'),
      ),
      pipe: call(params.pipe),
    };
  },
};
